src/box/memtx_sort_data.c

```diff
@@ -0,0 +1,711 @@
+		say_error("%s: invalid entry count", reader->filename);
+		goto fail_close;
+	}
+
+	/* Get the sort data entries. */
+	for (unsigned i = 0; i < entry_count; i++) {
+		struct memtx_sort_data_reader_entry entry;
+		long entry_data_length;
+		if (fscanf(reader->fp, ENTRY_FMT, &entry.key.space_id,
+			   &entry.key.index_id, &entry.offset,
+			   &entry.psize, &entry_data_length) != 5) {
+			say_error("%s: entry read failed", reader->filename);
+			goto fail_close;
+		}
+		/*
+		 * Simple early verification, will fail on tuple
+		 * translation later in case the file is truncated.
+		 */
+		if (entry.psize % entry_data_length != 0) {
+			say_error("%s: invalid entry length", reader->filename);
+			goto fail_close;
+		}
+		mh_memtx_sort_data_entries_put(reader->entries,
+					       &entry, NULL, NULL);
+	}
+
+	say_info("using the memtx sort data from `%s'", reader->filename);
+	return reader;
+
+fail_close:
+	fclose(reader->fp);
+	say_warn("memtx sort data file `%s' ignored", reader->filename);
+fail_free:
+	mh_memtx_sort_data_entries_delete(reader->entries);
+	mh_ptrptr_delete(reader->old2new);
+	free(reader->buffer);
+	free(reader);
+	return NULL;
+}
+
+/**
+ * Find the space in the sort data file and seek to the PK sort data if exists.
+ */
+static int
+memtx_sort_data_reader_pk_init(struct memtx_sort_data_reader *reader,
+			       uint32_t space_id)
+{
+	/* Check if have sort data for the space in the file. */
+	struct memtx_sort_data_key key = {space_id, 0};
+	mh_int_t i = mh_memtx_sort_data_entries_find(reader->entries,
+						     key, NULL);
+	if (i == mh_end(reader->entries)) {
+		reader->curr_entry = NULL;
+		return 0;
+	}
+
+	/* Seek to the PK data and save the entry info to access later. */
+	struct memtx_sort_data_reader_entry *entry =
+		mh_memtx_sort_data_entries_node(reader->entries, i);
+	if (fseek(reader->fp, entry->offset, SEEK_SET) != 0) {
+		DIAG_SET(reader, "space %u: PK seek failed", space_id);
+		return -1;
+	}
+	reader->curr_entry = entry;
+	return 0;
+}
+
+int
+memtx_sort_data_reader_pk_add_tuple(struct memtx_sort_data_reader *reader,
+				    uint32_t space_id, struct tuple *tuple)
+{
+	/* First tuple inserted into the space? */
+	if (reader->curr_space_id != space_id) {
+		if (memtx_sort_data_reader_pk_init(reader, space_id) != 0)
+			return -1;
+		reader->curr_space_id = space_id;
+	}
+
+	/* No sort data for the space? */
+	if (reader->curr_entry == NULL)
+		return 0;
+
+	/* Consistency check. */
+	assert(reader->curr_entry->key.space_id == space_id);
+
+	/*
+	 * Associate the old tuple pointer (read from the sort data file) with
+	 * the new one (created on insertion). The fread is buffered with user
+	 * specified buffer of a large size, see the reader constructor.
+	 */
+	struct mh_ptrptr_node_t node;
+	if (fread(&node.key, sizeof(node.key), 1, reader->fp) != 1) {
+		DIAG_SET(reader, "space %u: PK read failed", space_id);
+		return -1;
+	}
+	node.val = tuple;
+	mh_ptrptr_put(reader->old2new, &node, NULL, NULL);
```

Comment:
Do we really need to maintain the hash table for all spaces during the whole recovery procedure? Can't we build it for one space at a time? I mean, read one space, build its indexes, then read the next space, and so on.